libbpfgo.go

```diff
@@ -113,6 +113,28 @@ int bpf_prog_detach_cgroup_legacy(
 
 	return syscall(__NR_bpf, BPF_PROG_DETACH, &attr, sizeof(attr));
 }
+
+struct bpf_object * open_bpf_object(
+        char* btf_file_path,
+        char* kconfig_path,
+        char* bpf_obj_name,
+        const void* obj_buf,
+        size_t obj_buf_size)
+{
+	struct bpf_object_open_opts opts = {};
+	opts.btf_custom_path = btf_file_path;
+	opts.kconfig = kconfig_path;
+	opts.object_name = bpf_obj_name;
+	opts.sz = sizeof(opts);
+
+	struct bpf_object *obj = bpf_object__open_mem(obj_buf, obj_buf_size, &opts);
+	if (obj == NULL) {
+			fprintf(stderr, "Failed to open bpf object: %s\n", strerror(errno));
```

Comment:
I suppose that `fprintf()` in rare cases can change errno, what could make `C.open_bpf_object()` return a wrong one. WDYT?

Reply:
Are you sure about this? Wouldn't it be passed by value?

Reply:
I'm supposing based on these:

https://github.com/bminor/glibc/blob/master/stdio-common/vfprintf-internal.c#L681-L683
https://github.com/bminor/glibc/blob/master/stdio-common/vfprintf-internal.c#L714
https://github.com/bminor/glibc/blob/master/stdio-common/vfprintf-internal.c#L887
https://github.com/bminor/glibc/blob/master/stdio-common/vfprintf-internal.c#L914
https://github.com/bminor/glibc/blob/master/stdio-common/vfprintf-internal.c#L937
https://github.com/bminor/glibc/blob/master/stdio-common/vfprintf-internal.c#L960
https://github.com/bminor/glibc/blob/master/stdio-common/vfprintf-internal.c#L1075

Reply:
The thing is that `strerror()` has this in its manual:

```
       POSIX.1-2001 and POSIX.1-2008 require that a successful  call  to  str‐
       error()  or  strerror_l()  shall  leave errno unchanged, and note that,
       since no function return value is reserved to indicate an error, an ap‐
       plication  that  wishes  to check for errors should initialize errno to
       zero before the call, and then check errno after the call.
```

But none are mentioned about errno in `fprintf` manual. And besides saving current errno in `save_errno` I didn't find it returning it as an error. It returns done -1. I could be wrong for sure, just skimming on this matter.

Reply:
Try this:

```c
#include <stdio.h>
#include <errno.h>
#include <string.h>

int main(void) {
    printf("errno: %d - %s\n", errno, strerror(errno));
    fprintf(stderr, "TESTING: %'.5000000000f'\n", 1234567.89);
    printf("errno: %d - %s\n", errno, strerror(errno));

    return 0;
}
```

```bash
❯ ./test            
errno: 0 - Success
TESTING: errno: 75 - Value too large for defined data type
```

So, based on this I assume that printf family functions can taint errno. And of course, your current code is error free, but future changes can cause errors to arise. The same for `init_ring_buf` and `init_perf_buf`.

Reply:
```suggestion
			int saved_errno = errno;
			fprintf(stderr, "Failed to open bpf object: %s\n", strerror(errno));
			errno = saved_errno;
```

Reply:
So is the fix for this just this?

```
	struct bpf_object *obj = bpf_object__open_mem(obj_buf, obj_buf_size, &opts);
	if (obj == NULL) {
			int saved_errno = errno;
			fprintf(stderr, "Failed to open bpf object: %s\n", strerror(saved_errno));
			return NULL;
	}
```

Reply:
> Very interesting!! So is the fix for this just this?

I would:

 ```c
    if (obj == NULL) {
        int saved_errno = errno;
        fprintf(stderr, "Failed to open bpf object: %s\n", strerror(errno));
        errno = saved_errno;
        return NULL;
    }
 ```
So we keep the main (right) errno returning to go.

Reply:
Just an addendum: https://man7.org/linux/man-pages/man3/errno.3.html#NOTES

Reply:
Done, looks like the pattern is done in a couple other places too, i'll add them in a follow up PR.